lib/services/matchers.go

```diff
@@ -57,6 +58,124 @@ func MatchResourceLabels(matchers []ResourceMatcher, resource types.ResourceWith
 	return false
 }
 
+// MatchResourceByFilters returns true if all filter values given matched against the resource.
+func MatchResourceByFilters(resource types.ResourceWithLabels, filter MatchResourceFilter) (bool, error) {
+	if filter.Labels == nil && len(filter.PredicateExpression) == 0 && len(filter.SearchKeywords) == 0 {
+		return true, nil
+	}
+
+	var specResource types.ResourceWithLabels
+
+	// We assume when filtering for services like KubeService, AppServer, and DatabaseServer
+	// the user is wanting to filter the contained resource ie. KubeClusters, Application, and Database.
+	switch filter.ResourceKind {
+	case types.KindNode:
+		specResource = resource
+
+	case types.KindKubeService:
+		return matchAndFilterKubeClusters(resource, filter)
+
+	case types.KindAppServer:
+		server, ok := resource.(types.AppServer)
+		if !ok {
+			return false, trace.BadParameter("expected types.AppServer, got %T", resource)
+		}
+		specResource = server.GetApp()
+
+	case types.KindDatabaseServer:
+		server, ok := resource.(types.DatabaseServer)
+		if !ok {
+			return false, trace.BadParameter("expected types.DatabaseServer, got %T", resource)
+		}
+		specResource = server.GetDatabase()
+
+	default:
+		return false, trace.NotImplemented("filtering for resource kind %q not supported", filter.ResourceKind)
+	}
+
+	return matchResourceByFilters(specResource, filter)
+}
+
+func matchResourceByFilters(resource types.ResourceWithLabels, filter MatchResourceFilter) (bool, error) {
+	if len(filter.PredicateExpression) > 0 {
+		parser, err := NewResourceParser(resource)
+		if err != nil {
+			return false, trace.Wrap(err)
+		}
+
+		switch match, err := parser.EvalBoolPredicate(filter.PredicateExpression); {
+		case err != nil:
+			return false, trace.Wrap(err)
+		case !match:
+			return false, nil
+		}
+	}
+
+	if !types.MatchLabels(resource, filter.Labels) {
+		return false, nil
+	}
+
+	if !resource.MatchSearch(filter.SearchKeywords) {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// matchAndFilterKubeClusters is similar to MatchResourceByFilters, but does two things in addition:
+//  1) handles kube service having a 1-N relationship (service-clusters)
+//     so each kube cluster goes through the filters
+//  2) filters out the non-matched clusters on the kube service and the kube service is
+//     updated in place with only the matched clusters
```

Comment:
Maybe this should bit of documentation should be bubbled up to `MatchResourceByFilters` so we don't accidentally modify something we didn't mean to modify in the future.